src/trajectory/util.cpp

```diff
@@ -0,0 +1,314 @@
+  std::size_t currSegmentIdx = 0u;
+
+  const auto segmentStartState = stateSpace->createState();
+  const auto segmentEndState = stateSpace->createState();
+  Eigen::VectorXd segStartPos(dimension);
+  Eigen::VectorXd segEndPos(dimension);
+  Eigen::VectorXd segStartVel(dimension);
+  Eigen::VectorXd segEndVel(dimension);
+  const Eigen::VectorXd zeroPos = Eigen::VectorXd::Zero(dimension);
+  traj.evaluate(partialStartTime, segmentStartState);
+  stateSpace->logMap(segmentStartState, segStartPos);
+  traj.evaluateDerivative(partialStartTime, 1, segStartVel);
+
+  // Find a segment of original trajectory that includes partialStartTime
+  while (currSegmentIdx < traj.getNumSegments())
+  {
+    currSegmentEndTime += traj.getSegmentDuration(currSegmentIdx);
+    if (currSegmentStartTime <= partialStartTime
+        && partialStartTime <= currSegmentEndTime)
+    {
+      // create new segment
+      traj.evaluate(currSegmentEndTime, segmentEndState);
+      stateSpace->logMap(segmentEndState, segEndPos);
+      traj.evaluateDerivative(currSegmentEndTime, 1, segEndVel);
+
+      const double segmentDuration = currSegmentEndTime - partialStartTime;
+
+      if (segmentDuration > 0.0)
+      {
+        CubicSplineProblem problem(Vector2d{0., segmentDuration}, 4, dimension);
+        problem.addConstantConstraint(0, 0, zeroPos);
+        problem.addConstantConstraint(0, 1, segStartVel);
+        problem.addConstantConstraint(1, 0, segEndPos - segStartPos);
+        problem.addConstantConstraint(1, 1, segEndVel);
+        const auto solution = problem.fit();
+        const auto coefficients = solution.getCoefficients().front();
+        outputTrajectory->addSegment(
+            coefficients, segmentDuration, segmentStartState);
+      }
+      break;
+    }
+
+    currSegmentIdx++;
+    currSegmentStartTime = currSegmentEndTime;
+  }
+
+  for (std::size_t i = currSegmentIdx + 1u; i < traj.getNumSegments(); i++)
+  {
+    outputTrajectory->addSegment(
+        traj.getSegmentCoefficients(i),
+        traj.getSegmentDuration(i),
+        traj.getSegmentState(i));
+  }
+
+  return outputTrajectory;
+}
+
+//==============================================================================
+UniqueInterpolatedPtr convertToInterpolated(
+    const Spline& traj, statespace::ConstInterpolatorPtr interpolator)
+{
+  auto stateSpace = traj.getStateSpace();
+  auto outputTrajectory
+      = make_unique<Interpolated>(stateSpace, std::move(interpolator));
+
+  auto state = stateSpace->createState();
+  for (std::size_t i = 0; i < traj.getNumWaypoints(); i++)
+  {
+    traj.getWaypoint(i, state);
+    const double t = traj.getWaypointTime(i);
+    outputTrajectory->addWaypoint(t, state);
+  }
+
+  return outputTrajectory;
+}
+
+//==============================================================================
+UniqueInterpolatedPtr concatenate(
+    const Interpolated& traj1, const Interpolated& traj2)
+{
+  if (traj1.getStateSpace()->getDimension()
```

Comment:
Should we check that the trajectories' state spaces and interpolators are the same instance, rather than just dimension?